File: Raven.Client.Lightweight/Document/HiLoKeyGeneratorBase.cs
using System;
using System.Linq;
using System.Threading;
using Raven.Abstractions;
using Raven.Abstractions.Data;
using Raven.Abstractions.Util;
using Raven.Client.Connection;

namespace Raven.Client.Document
{
    public abstract class HiLoKeyGeneratorBase
    {
        protected const string RavenKeyGeneratorsHilo = "Raven/Hilo/";
        protected const string RavenKeyServerPrefix = "Raven/ServerPrefixForHilo";

        protected readonly string tag;
        protected long capacity;
        protected long baseCapacity;
        private volatile RangeValue range;

        protected string lastServerPrefix;
        protected DateTime lastRequestedUtc1, lastRequestedUtc2;

        protected readonly ManualResetEvent mre = new ManualResetEvent(false);
        protected InterlockedLock interlockedLock = new InterlockedLock();
        protected long threadsWaitingForRangeUpdate = 0;

        protected HiLoKeyGeneratorBase(string tag, long capacity)
        {
            this.tag = tag;
            this.capacity = capacity;
            baseCapacity = capacity;
            this.range = new RangeValue(1, 0);
        }

        protected string GetDocumentKeyFromId(DocumentConvention convention, long nextId)
        {
            return string.Format("{0}{1}{2}{3}",
                                 tag,
                                 convention.IdentityPartsSeparator,
                                 lastServerPrefix,
                                 nextId);
        }

        protected long GetMaxFromDocument(JsonDocument document, long minMax, long calculatedCapacity)
        {
            long max;
            if (document.DataAsJson.ContainsKey("ServerHi")) // convert from hi to max
            {
                var hi = document.DataAsJson.Value<long>("ServerHi");
                max = ((hi - 1) * calculatedCapacity);
                document.DataAsJson.Remove("ServerHi");
                document.DataAsJson["Max"] = max;
            }
            max = document.DataAsJson.Value<long>("Max");
            return Math.Max(max, minMax);
        }

        protected string HiLoDocumentKey
        {
            get { return RavenKeyGeneratorsHilo + tag; }
        }

        public bool DisableCapacityChanges { get; set; }

        protected void ModifyCapacityIfRequired(ref long calculatedCapacity)
        {
            if (DisableCapacityChanges)
                return;

            var capacityChanged = false;
            var span = SystemTime.UtcNow - lastRequestedUtc1;
            if (span.TotalSeconds < 5)
            {
                capacityChanged = true;
                span = SystemTime.UtcNow - lastRequestedUtc2;
                if (span.TotalSeconds < 3)
                    calculatedCapacity = Math.Max(calculatedCapacity, Math.Max(calculatedCapacity * 2, calculatedCapacity * 4));
                else
                    calculatedCapacity = Math.Max(calculatedCapacity, calculatedCapacity * 2);
            }
            else if (span.TotalMinutes > 1)
            {
                capacityChanged = true;
                calculatedCapacity = Math.Max(baseCapacity, calculatedCapacity / 2);
            }

            if (capacityChanged)
            {
                //there is only one thread that is updating the capacity
                capacity = calculatedCapacity;
            }
                
            lastRequestedUtc2 = lastRequestedUtc1;
            lastRequestedUtc1 = SystemTime.UtcNow;
        }

        protected void IncreaseCapacityIfRequired(ref long calculatedCapacity)
        {
            var waitingForRangeUpdate = Interlocked.Read(ref threadsWaitingForRangeUpdate);
            var capacityChanged = false;

            //we want to have at least 90% of free capacity in the new range
            while ((calculatedCapacity - waitingForRangeUpdate) < 0.9 * calculatedCapacity)
            {
                capacityChanged = true;
                calculatedCapacity = calculatedCapacity * 2;
            }

            if (capacityChanged)
            {
                //there is only one thread that is updating the capacity
                capacity = calculatedCapacity;
            }
        }

        protected JsonDocument HandleGetDocumentResult(MultiLoadResult documents)
        {
            if (documents.Results.Count == 2 && documents.Results[1] != null)
            {
                lastServerPrefix = documents.Results[1].Value<string>("ServerPrefix");
            }
            else
            {
                lastServerPrefix = string.Empty;
            }
            if (documents.Results.Count == 0 || documents.Results[0] == null)
                return null;
            var jsonDocument = documents.Results[0].ToJsonDocument();
            foreach (var key in jsonDocument.Metadata.Keys.Where(x => x.StartsWith("@")).ToArray())
            {
                jsonDocument.Metadata.Remove(key);
            }
            return jsonDocument;
        }

        protected RangeValue Range
        {
            get { return range; }
            set { range = value; }
        }

        [System.Diagnostics.DebuggerDisplay("[{Min}-{Max}]: {Current}")]
        protected class RangeValue
        {
            public readonly long Min;
            public readonly long Max;
            public long Current;

            public RangeValue(long min, long max)
            {
                this.Min = min;
                this.Max = max;
                this.Current = min - 1;
            }
        }
    }
}

File: Raven.Client.Java/src/main/java/net/ravendb/client/document/HiLoKeyGeneratorBase.java
package net.ravendb.client.document;

import java.util.Date;
import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.atomic.AtomicLong;

import net.ravendb.abstractions.data.JsonDocument;
import net.ravendb.abstractions.data.MultiLoadResult;
import net.ravendb.abstractions.json.linq.RavenJValue;
import net.ravendb.client.connection.SerializationHelper;


public abstract class HiLoKeyGeneratorBase {

  protected final String RAVEN_KEY_GENERATORS_HILO = "Raven/Hilo/";
  protected final String RAVEN_KEY_SERVER_PREFIX = "Raven/ServerPrefixForHilo";

  protected final String tag;
  protected long capacity;
  protected long baseCapacity;
  private volatile RangeValue range;

  protected String lastServerPrefix;
  protected long lastRequestedUtc1, lastRequestedUtc2;

  private boolean disableCapacityChanges;

  protected HiLoKeyGeneratorBase(String tag, long capacity) {
    this.tag = tag;
    this.capacity = capacity;
    baseCapacity = capacity;
    this.range = new RangeValue(1, 0);
  }

  protected String getDocumentKeyFromId(DocumentConvention convention, long nextId) {
    return tag + convention.getIdentityPartsSeparator()+lastServerPrefix+nextId;
  }

  protected long getMaxFromDocument(JsonDocument document, long minMax) {
    long max;
    if (document.getDataAsJson().containsKey("ServerHi")) { // convert from hi to max
      Long hi = document.getDataAsJson().value(Long.TYPE, "ServerHi");
      max = ((hi - 1) * capacity);
      document.getDataAsJson().remove("ServerHi");
      document.getDataAsJson().add("Max", new RavenJValue(max));
    }
    max = document.getDataAsJson().value(Long.TYPE, "Max");
    return Math.max(max, minMax);
  }

  protected String getHiLoDocumentKey() {
    return RAVEN_KEY_GENERATORS_HILO + tag;
  }

  public boolean isDisableCapacityChanges() {
    return disableCapacityChanges;
  }

  public void setDisableCapacityChanges(boolean disableCapacityChanges) {
    this.disableCapacityChanges = disableCapacityChanges;
  }

  protected void modifyCapacityIfRequired() {
    if (disableCapacityChanges) {
      return;
    }

    long span = new Date().getTime() - lastRequestedUtc1;
    if (span < 5 * 1000) {
      span = new Date().getTime() - lastRequestedUtc2;
      if (span < 3000) {
        capacity *= 4;
      } else {
        capacity *= 2;
      }
    } else if (span > 60 * 1000) {
      capacity = Math.max(baseCapacity, capacity / 2);
    }

    lastRequestedUtc2 = lastRequestedUtc1;
    lastRequestedUtc1 = new Date().getTime();
  }

  protected JsonDocument handleGetDocumentResult(MultiLoadResult documents) {
    if (documents.getResults().size() == 2 && documents.getResults().get(1) != null) {
      lastServerPrefix = documents.getResults().get(1).value(String.class, "ServerPrefix");
    } else {
      lastServerPrefix = "";
    }
    if (documents.getResults().isEmpty() || documents.getResults().get(0) == null) {
      return null;
    }
    JsonDocument jsonDocument = SerializationHelper.toJsonDocument(documents.getResults().get(0));

    Set<String> metaKeysToRemove = new HashSet<>();
    for (String key: jsonDocument.getMetadata().getKeys()) {
      if (key.startsWith("@")) {
        metaKeysToRemove.add(key);
      }
    }
    for (String r: metaKeysToRemove) {
      jsonDocument.getMetadata().remove(r);
    }
    return jsonDocument;
  }


  protected RangeValue getRange() {
    return range;
  }

  protected void setRange(RangeValue range) {
    this.range = range;
  }


  protected static class RangeValue {
    public AtomicLong min;
    public AtomicLong max;
    public AtomicLong current;

    public RangeValue(long min, long max)
    {
      this.min = new AtomicLong(min);
      this.max = new AtomicLong(max);
      this.current = new AtomicLong(min - 1);
    }
  }
}
